velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,286 @@
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      vector_size_t resultOffset,
+      const VectorPtr& result) override {
+    int numRows = frameStarts->size() / sizeof(vector_size_t);
+    auto frameStartsVector = frameStarts->as<vector_size_t>();
+    auto frameEndsVector = frameEnds->as<vector_size_t>();
+
+    FrameMetadata frameMetadata =
+        fillArgVectors(frameStartsVector, frameEndsVector, numRows);
+    if (frameMetadata.fixedFrameStart && frameMetadata.increasingFrame) {
+      fixedStartAggregation(
+          numRows,
+          frameMetadata.minFrame,
+          frameMetadata.maxFrame,
+          frameEndsVector,
+          resultOffset,
+          result);
+    } else {
+      naiveAggregation(
+          numRows,
+          frameMetadata.minFrame,
+          frameMetadata.maxFrame,
+          frameStartsVector,
+          frameEndsVector,
+          resultOffset,
+          result);
+    }
+  }
+
+ private:
+  struct FrameMetadata {
+    vector_size_t minFrame;
+    vector_size_t maxFrame;
+    bool fixedFrameStart;
+    bool increasingFrame;
+  };
+
+  FrameMetadata fillArgVectors(
+      const vector_size_t* frameStarts,
+      const vector_size_t* frameEnds,
+      vector_size_t numRows) {
+    // Argument values are needed to compute the aggregate. The aggregate
+    // computation for each row needs the values from its frameStart to frameEnd
+    // indices. So for each apply() we read arguments from the least
+    // frameStart to the highest frameEnd value in the row block.
+    vector_size_t minFrame = frameStarts[0];
+    vector_size_t maxFrame = frameEnds[0];
+
+    bool increasingFrame = true;
+    for (int i = 1; i < numRows; i++) {
+      minFrame = std::min(minFrame, frameStarts[i]);
+      maxFrame = std::max(maxFrame, frameEnds[i]);
+      increasingFrame &= frameEnds[i] >= frameEnds[i - 1];
+    }
+    bool fixedMinFrame = minFrame == frameStarts[0];
+
+    vector_size_t numFrameRows = maxFrame + 1 - minFrame;
+    for (int i = 0; i < argIndices_.size(); i++) {
+      argVectors_[i]->resize(numFrameRows);
+      partition_->extractColumn(
+          argIndices_[i], minFrame, numFrameRows, 0, argVectors_[i]);
+    }
+
+    return {minFrame, maxFrame, fixedMinFrame, increasingFrame};
+  }
+
+  void fixedStartAggregation(
+      vector_size_t numRows,
+      vector_size_t minFrame,
+      vector_size_t maxFrame,
+      const vector_size_t* frameEndsVector,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    SelectivityVector rows;
+    rows.resize(maxFrame + 1 - minFrame);
+
+    auto singleGroup = std::vector<vector_size_t>{0};
+    // Initialize the aggregate at the beginning of the output block.
+    aggregate_->clear();
+    aggregate_->initializeNewGroups(&singleGroupRow_, singleGroup);
+    auto frameBeginIndex = minFrame;
```

Comment:
Variable names are confusing to me.

frameBeginIndex -> prevFrameEnd or prevFrameSize?

Is the initial value here correct? Should it be 0?